src/basic/mount-util.c

```diff
@@ -874,3 +878,77 @@ int mount_propagation_flags_from_string(const char *name, unsigned long *ret) {
                 return -EINVAL;
         return 0;
 }
+
+int mount_option_mangle(const char *options, unsigned long *ret_mount_flags, char **ret_remaining_options) {
+        const struct libmnt_optmap *map;
+        _cleanup_free_ char *ret = NULL;
+        unsigned long mount_flags;
+        const char *p;
+        int r;
+
+        /* This extracts mount flags from the mount options, and store
+         * non-mount-flag options to *ret_remaining_options.
+         * E.g.,
+         * "rw,nosuid,nodev,relatime,size=1630748k,mode=700,uid=1000,gid=1000"
+         * is split to MS_NOSUID|MS_NODEV|MS_RELATIME and
+         * "size=1630748k,mode=700,uid=1000,gid=1000".
+         * See more examples in test-mount-utils.c.
+         *
+         * Note that *ret_mount_flags is not initialized in this function.
+         * The caller needs to initialized to zero or some default vaule. */
+
+        assert(options);
+        assert(ret_mount_flags);
+        assert(ret_remaining_options);
+
+        mount_flags = *ret_mount_flags;
+        ret = strdup("");
+        if (!ret)
+                return -ENOMEM;
+
+        map = mnt_get_builtin_optmap(MNT_LINUX_MAP);
+        if (!map)
+                return -EINVAL;
+
+        p = options;
+        for (;;) {
+                _cleanup_free_ char *word = NULL;
+                const struct libmnt_optmap *ent;
+
+                r = extract_first_word(&p, &word, ",", EXTRACT_QUOTES);
+                if (r < 0)
+                        return r;
+                if (r == 0)
+                        break;
+
+                for (ent = map; ent->name; ent++) {
+                        /* All entries in MNT_LINUX_MAP do not take any argument.
+                         * Thus, ent->name does not contain "=" or "[=]". */
+                        if (!streq(word, ent->name))
+                                continue;
+
+                        if (!(ent->mask & MNT_INVERT))
+                                mount_flags |= ent->id;
+                        else if (mount_flags & ent->id)
+                                mount_flags ^= ent->id;
+
+                        break;
+                }
+
+                if (!ent->name) {
+                        char *t;
+
+                        t = strjoin(ret, isempty(ret) ? "" : ",", word);
```

Comment:
looks like a perfect job for strextend()

Reply:
`strextend()` also has the nice benefit that you can drop the `strdup("")` call further up, as it handles NULL strings as first parameter just fine. That way we can save one allocation in the common case, and then do the strdup("") only when ret is still NULL at the end...